CHANGES.md

```diff
@@ -1,3 +1,58 @@
+### Unreleased
+
+- This release introduces a significant change in the Mirage tool by
+  splitting the configure-time and runtime keys. Configure-time keys
+  are essential during the setup of module dependencies for the
+  unikernel, allowing for a specialized production of a unikernel for
+  a given target runtime environment. On the other hand, runtime keys
+  are useful for customizing deployments without altering the
+  dependencies of the unikernels. (#1449, #1450, #1451, #1455 @samoht,
+  review by @hannesm)
+
+  * API changes:
+    - There is no more `~stage` parameter for `Mirage.Key.Arg.info`.
+    - `Mirage.Key` now define configure-time keys only.
+    - There is a new module `Mirage.Runtime_key` to define runtime keys.
+    - As there are no more keys type ``Both`, users are now expected to create
+      two separated keys in that case (one for configure-time, one for runtime)
+      or decide if the key is useful at runtime of configure-time.
+
+  * Intended use of configure-time keys:
+    - Used to set up module dependencies of the unikernel, such as the
+      target (hvt, xen, etc.) and whether to use DHCP or a fixed IP address.
+    - Enable the production of specialized unikernels suitable for
+      specific target runtime environments and dedicated network and
+      storage stacks.
+    - Similar keys will produce reproducible binaries to be uploaded to artifact
+      repositories like Docker Hub or https://builds.robur.coop/.
+
+  * Intended use of runtime keys:
+    - Allow users to customize deployments, like IP addresses, secrets,
+      block device names, etc., post downloading of binaries.
+    - These keys don’t alter the dependencies of the unikernels.
+    - A runtime keys in the Mirage DSL is just a reference to a normal Cmdliner
+      term.
+
+  * Code migration:
+    In `config.ml`:
+    ```ocaml
+    let key =
+      let doc = Key.Arg.info ~doc:"A Key." ~stage:`Run [ "key" ] in
+      Key.(create "key" Arg.(opt_all ~stage:`Run string doc))
+    ```
+    becomes:
+    ```ocaml
+    let key = Runtime_key.create "Unikernel.key"
+    ```
+    and `unikenel.ml` now contains:
+    ```ocaml
+    open Cmdliner
+
+    let key =
+      let doc = Arg.info ~doc:"A Key." [ "key" ] in
+      Arg.(value & opt_all string [] doc)
+    ```
```

Comment:
can we now as well define runtime keys purely in unikernel.ml (or in any library that depends on mirage-runtime)? or is there a requirement to have runtime keys of a unikernel mentioned in config.ml (if yes, why)?

Reply:
Ha yes you are right I will clarify this. It's not a requirement to define the runtime key in `config.ml` - you do this only if you want to refer it in some new device initialisation code. I'll clarify.